File: ts/src/ollama/options.ts
import type { RunOptions, OnMessageFn, Logger } from "../types.js";

/**
 * Function that performs an HTTP fetch. Used for dependency injection in tests.
 * Must accept the same signature as the global fetch.
 */
export type FetchFn = typeof fetch;

/** Configuration for creating an Ollama runner. */
export interface OllamaRunnerConfig {
  /** Base URL for the Ollama API (default: "http://localhost:11434"). */
  baseURL?: string;

  /** Inject a custom fetch function for testing. */
  fetch?: FetchFn;

  /** Structured logger for debug output (nil by default). */
  logger?: Logger;
}

/** Ollama-specific options that extend the common RunOptions. */
export interface OllamaRunOptions extends RunOptions {
  /** Sampling temperature. */
  temperature?: number;

  /** Context window size. */
  numCtx?: number;

  /** Maximum number of tokens to generate. */
  numPredict?: number;

  /** Random seed for reproducible generation. */
  seed?: number;

  /** Stop sequences. */
  stop?: string[];

  /** Top-k sampling parameter. */
  topK?: number;

  /** Top-p (nucleus) sampling parameter. */
  topP?: number;

  /** Min-p sampling parameter. */
  minP?: number;

  /** Response format (e.g. "json"). */
  format?: string;

  /** How long the model stays loaded (e.g. "5m", "0" to unload immediately). */
  keepAlive?: string;

  /** Enable thinking/reasoning for supported models (e.g. qwen3). */
  think?: boolean;

  /** Callback invoked for each streaming message. */
  onMessage?: OnMessageFn;
}

File: python/src/agentrunner/ollama/options.py
"""Ollama runner configuration and run options."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Protocol, runtime_checkable

from ..types import RunOptions


@runtime_checkable
class Logger(Protocol):
    """Minimal logger interface (matches stdlib logging.Logger)."""

    def debug(self, msg: str, *args: Any, **kwargs: Any) -> None: ...
    def error(self, msg: str, *args: Any, **kwargs: Any) -> None: ...


@dataclass
class OllamaRunnerConfig:
    """Configuration for creating an Ollama runner."""

    base_url: str = "http://localhost:11434"
    logger: Logger | None = None


@dataclass
class OllamaRunOptions(RunOptions):
    """Ollama-specific options extending common RunOptions."""

    temperature: float | None = None
    num_ctx: int | None = None
    num_predict: int | None = None
    seed: int | None = None
    stop: list[str] | None = field(default=None)
    top_k: int | None = None
    top_p: float | None = None
    min_p: float | None = None
    format: str | None = None
    keep_alive: str | None = None
    think: bool | None = None
